noxfile.py

```diff
@@ -0,0 +1,280 @@
+        args += ["--", "gcovr", "doc/examples"]
+    session.run("python", *args)
+
+
+@nox.session
+def build_wheel(session: nox.session) -> None:
+    """Build a wheel."""
+    session.install("wheel", "twine")
+    session.run("python", "setup.py", "sdist", "bdist_wheel")
+    session.run("twine", "check", "dist/*", external=True)
+
+
+@nox.session
+def upload_wheel(session: nox.session) -> None:
+    """Upload the wheel."""
+    session.install("twine")
+    session.run("twine", "upload", "dist/*", external=True)
+
+
+def docker_container_id(version: str) -> str:
+    """Get the docker container ID."""
+    return f"gcovr-qa-{version}-uid_{os.geteuid()}"
+
+
+@nox.session(python=False)
+def docker_qa_build(session: nox.session) -> None:
+    """Build the docker container for the default GCC version."""
+    session_id = f"docker_qa_build_compiler({GCC_VERSION2USE})"
+    session.log(f"Notify session {session_id}")
+    session.notify(session_id)
+
+
+@nox.session(python=False)
+def docker_qa_build_all_compiler(session: nox.session) -> None:
+    """Build the docker containers vor all GCC versions."""
+    for version in GCC_VERSIONS:
+        session_id = f"docker_qa_build_compiler({version})"
+        session.log(f"Notify session {session_id}")
+        session.notify(session_id)
+
+
+@nox.session(python=False)
+@nox.parametrize("version", [nox.param(v, id=v) for v in GCC_VERSIONS])
+def docker_qa_build_compiler(session: nox.session, version: str) -> None:
+    """Build the docker container for a specific GCC version."""
+    set_environment(session, version, False)
+    session.run(
+        "docker",
+        "build",
+        "--tag",
+        docker_container_id(version),
+        "--build-arg",
+        f"USERID={os.geteuid()}",
+        "--build-arg",
+        f"CC={session.env['CC']}",
+        "--build-arg",
+        f"CXX={session.env['CXX']}",
+        "--file",
+        "admin/Dockerfile.qa",
+        ".",
+        external=True,
+    )
+
+
+@nox.session(python=False)
+def docker_qa_run(session: nox.session) -> None:
+    """Run the docker container for the default GCC version."""
+    session_id = f"docker_qa_run_compiler({GCC_VERSION2USE})"
+    session.log(f"Notify session {session_id}")
+    session.notify(session_id)
+
+
+@nox.session(python=False)
+def docker_qa_run_all_compiler(session: nox.session) -> None:
+    """Run the docker container for the all GCC versions."""
+    for version in GCC_VERSIONS:
+        session_id = f"docker_qa_run_compiler({version})"
+        session.log(f"Notify session {session_id}")
+        session.notify(session_id)
+
+
+@nox.session(python=False)
+@nox.parametrize("version", [nox.param(v, id=v) for v in GCC_VERSIONS])
+def docker_qa_run_compiler(session: nox.session, version: str) -> None:
+    """Run the docker container for a specific GCC version."""
+    set_environment(session, version, False)
+    session.env["NOX_POSARGS"] = " ".join([repr(a) for a in session.posargs])
```

Comment:
This variable is later expanded via `shlex.split()`. Then, we should probably use `shlex.join(session.posargs)` here as well.

Reply:
I'll change this.

Reply:
`shlex.join` was introduced in python 3.8. I'll use this if available, else a copy of the function is used.